### PR TITLE
Add user filtering to the Notebook

### PR DESCRIPTION
**Pull request by lyzadanger**
This is a ~~draft~~ PR of all the bits and pieces needed to add rudimentary user filtering to the Notebook. What I'd like to ascertain with this ~~draft~~ PR is:

- [x] General approach OK?
- [x] Would reviewer preference be to break up this into smaller PRs? Each commit here is self-sufficient, so it can be chunked up in arbitrary ways. (Decided: keep as one PR)

## A Few Screenshots

On launch (not in user-focus mode), Notebook looks like this:

image

When filtering by user, the count of matches changes to use a "results" unit:

image

"Force-expanding" a non-matching thread will update the results count similar to in the sidebar:

image

When user-focus mode is active (e.g. in grading mode), opening the Notebook will filter by the currently-focused user initially:

image

(but the user can change which user to filter by, or set it to "Everybody").

## A List

You should expect these changes to accomplish the following:

* A user filter should appear in the notebook. This simple filter shows one user per "line"
* On initial opening, the user filter should indicate that “Everybody” ’s annotations are showing
* The user filter uses the "profile" icon to indicate its user-y-ness (but also has an accessible title)
* The user menu should include all users who have at least one annotation in the current group
* Users should be in alphabetical order
	* If the `client_display_names` feature flag is enabled, the user’s display names will display, and will be alphabetized
	* If not, the names shown will be (alphabetized) usernames
* Selecting a user from the filter menu should filter the annotations shown to annotations by that user
	* The count of matches should change from `X threads (y annotations)` to `X results` (or `1 result` if singular)
	* The user selected should be represented in the filter label
	* Opening the filter again should show the user visibly selected
	* Clicking on any “Show x more in conversation” buttons will update the match count to `X results (and y more)` similarly to how the sidebar handles this state
* Selecting “Everybody” from the user filter should show all annotations in the group again
* When in a user-focused mode:
	* On Notebook launch, that focused user will be selected in the user filter and their annotations shown
	* The focused user will show up in the user filter even if that user has no annotations in the current group
	* Other users, or “Everybody” can be selected as above to change the user filtering

### What's not included

Literally everything else, including things like URL hash/history management, handling really long lists of users, handling empty result sets more gracefully, anything but the most elementary visual styling, etc., etc.

_Edited_: Moving out of draft now

Part of https://github.com/hypothesis/client/issues/2731

### CODE REVIEW

File: src/sidebar/components/hooks/use-user-filter-options.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        ? username(annotation.user)
+        : annotation.user;
+      const displayValue =
+        showDisplayNames && annotation.user_info?.display_name
+          ? annotation.user_info.display_name
+          : _username;
+      users[annotation.user] = displayValue;
+    });
+
+    // If user-focus is configured (even if not applied) add a filter
```
Comment (lyzadanger) [inline src/sidebar/components/hooks/use-user-filter-options.js:36]: The comments in here were fantastically hard to write and I think it's still really confusing. This is both because my face is really close to the forest and because the underlying situation is smelly.

File: src/sidebar/components/notebook-results.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ * - Annotations are filtered: "X results [(and Y more)]"
+ */
+function NotebookResults() {
+  const store = useStoreProxy();
+  const forcedVisibleCount = store.forcedVisibleAnnotations().length;
+  const hasAppliedFilter = store.hasAppliedFilter();
+  const rootThread = useRootThread();
+  const visibleCount = countVisible(rootThread);
+
+  // TODO: Test that this is valid when in focused-user mode
```
Comment (lyzadanger) [inline src/sidebar/components/notebook-results.js:22]: Oops. This TODO is satisfied.

File: src/sidebar/components/test/notebook-view-test.js [hunk 3]
```diff
@@ -64,34 +64,8 @@ describe('NotebookView', () => {
...
-        fakeUseRootThread.returns(test.rootThread);
-        const wrapper = createComponent();
-
-        assert.equal(
-          wrapper.find('.notebook-view__results').text(),
-          test.expected
-        );
-      });
-    });
+  it('renders results (counts)', () => {
```
Comment (lyzadanger) [inline src/sidebar/components/test/notebook-view-test.js:67]: Looks like I missed a test for rendering filters, if we care.

File: src/sidebar/store/modules/filters.js [hunk 4]
```diff
@@ -169,10 +169,21 @@ function changeFocusModeUser(user) {
  * @param {FilterOption} filterOption
  */
 function setFilter(filterName, filterOption) {
-  return {
-    type: actions.SET_FILTER,
-    filterName,
-    filterOption,
+  return (dispatch, getState) => {
```
Comment (lyzadanger) [inline src/sidebar/store/modules/filters.js:172]: This is the other spot where the comments seem lame, and the underlying situation even lamer. Can discuss in person if it helps.

File: src/sidebar/components/hooks/test/use-user-filter-options-test.js [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      allAnnotations: sinon.stub().returns([]),
+      getFocusFilters: sinon.stub().returns({}),
+      isFeatureEnabled: sinon.stub().returns(false),
+    };
+
+    $imports.$mock({
+      '../../store/use-store': { useStoreProxy: () => fakeStore },
+    });
+
+    //mount(<DummyComponent />);
```
Comment (lyzadanger) [inline src/sidebar/components/hooks/test/use-user-filter-options-test.js:42]: Artifact of test refactor, whoops.

File: src/sidebar/components/test/select-filter-test.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    assert.calledWith(
+      fakeOnSelect,
+      sinon.match({ value: 'onevalue', display: 'One Value' })
+    );
+  });
+
+  it('should render provided icon and selected option in label', () => {
+    const wrapper = createComponent({ icon: 'profile' });
+
+    const label = mount(wrapper.find('Menu').props().label);
```
Comment (lyzadanger) [inline src/sidebar/components/test/select-filter-test.js:66]: These tests are where I ended up `mount`-ing a buried preact element...

**Comment by lyzadanger** [comment]
@robertknight I've removed one commit from this PR—the "fix" to the `totalChildren` property in the `build-thread` module, and am just using `rootThread.children.length`. Diff is slightly smaller now.

File: src/sidebar/components/hooks/use-user-filter-options.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+/** @typedef {import('../../store/modules/filters').FilterOption} FilterOption */
+
+/**
+ * Generate a list of users for filtering annotations; update when set of
+ * annotations or filter state changes meaningfully.
+ *
+ * @return {FilterOption[]}
+ */
+export default function useUserFilterOptions() {
```
Comment (lyzadanger) [inline src/sidebar/components/hooks/use-user-filter-options.js:14]: I'd like some feedback on whether a hook is the best way to encapsulate this logic. We need to think about: How would this adapt/be replaced if we relied on a (backend) service to provide the options for this filter facet, e.g.?

File: src/sidebar/components/hooks/use-user-filter-options.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        );
+      });
+
+      // Remove any pre-existing filter option corresponding to this user
+      // before (re-)adding it. The value defined by the focused user config
+      // needs to take precedence. This is so that if user focus is active,
+      // a filter component will correctly recognize the focused user as selected.
+      // Not doing this can result in two filter options for the same user:
+      // one using a username and one using an account ID.
+      if (alreadyPresent) {
```
Comment (lyzadanger) [inline src/sidebar/components/hooks/use-user-filter-options.js:59]: I suppose this conditional is unnecessary.

File: src/sidebar/components/hooks/use-user-filter-options.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    // option for that user. Note that this always respects the display
+    // value, even if `client_display_names` feature flags is not enabled:
+    // this matches current implementation of focus mode.
+    if (focusFilters.user) {
+      // We need to see if the user indicated by the user-focus filter is
+      // already contained in `users`. This is complicated by the fact that
+      // the value of the user filter can be either an account ID or a username;
+      // we can't be sure...
+      const maybeUsername = focusFilters.user.value;
+      const alreadyPresent = Object.keys(users).find(probablyAccountId => {
```
Comment (lyzadanger) [inline src/sidebar/components/hooks/use-user-filter-options.js:46]: `alreadyPresent` may be a poor naming choice here as it sounds boolean.

File: src/sidebar/components/select-filter.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+ * @prop {string} title
+ */
+
+/**
+ * A select-element-like control for selecting one of a defined set of
+ * options.
+ *
+ * @param {SelectFilterProps} props
+ */
+function SelectFilter({
```
Comment (lyzadanger) [inline src/sidebar/components/select-filter.js:28]: @robertknight You had asked at one point what we can do when there are too many filter options to make a select-list-style component feasible. I'm hoping that by managing UI in a component here, we can adapt, like, here. Perhaps we could set a threshold over which we convert to a text `input` with an `AutocompleteList` or similar. We could also go the _other_ way and render as radio buttons if there were few enough options (this is a common treatment for facets but requires some vertical space, which we'll have if we start laying out facets in the left of the grid on wider viewports. Yep, lots to think through still, but want to make sure we are putting the foundations in place...).

File: src/sidebar/components/select-filter.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+  const menuLabel = (
+    <span className="select-filter__menu-label">
+      {icon && <SvgIcon name={icon} className="select-filter__menu-icon" />}
+      {selected.display}
+    </span>
+  );
+
+  return (
+    <Menu label={menuLabel} title={title}>
```
Comment (lyzadanger) [inline src/sidebar/components/select-filter.js:47]: One thing this PR lacks is a11y tests. It's on my list. I may submit a separate PR with just that. I want to be absolutely certain that using a Menu component here is accessible. Again, by encapsulating this in a single `SelectFilter` component, I'm hoping it will be easy to swap out if we decide against it.

File: src/sidebar/components/hooks/test/use-user-filter-options-test.js [hunk 12]
```diff
@@ -0,0 +1,149 @@
+import { mount } from 'enzyme';
+import { createElement } from 'preact';
+
+import useUserFilterOptions from '../use-user-filter-options';
```
Comment (robertknight) [inline src/sidebar/components/hooks/test/use-user-filter-options-test.js:4]: I've been coming around to the view recently that default exports should be avoided in most cases. The primary reason is because it tends to increase the chance that a function/class is unnecessarily given different names in different files. On that basis I've recently avoided it except for UI components where there is a strong convention for the association of file and class names.

File: src/sidebar/components/hooks/test/use-user-filter-options-test.js [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+  afterEach(() => {
+    $imports.$restore();
+  });
+
+  it('should return a user filter option for each user who has authored an annotation', () => {
+    fakeStore.allAnnotations.returns(annotationFixtures());
+
+    mount(<DummyComponent />);
+
+    const values = lastUserOptions.map(option => option.value);
```
Comment (robertknight) [inline src/sidebar/components/hooks/test/use-user-filter-options-test.js:54]: You could just use one assert here and skip the `values` and `display` extraction. The same applies to the tests below.

```js
assert.deepEqual(lastUserOptions, [{ value: ..., display: ... }, { value: ..., display: ...}, ...])
```

File: src/sidebar/components/hooks/test/use-user-filter-options-test.js [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    ]);
+    assert.deepEqual(displays, [
+      'abalone',
+      'bananagram',
+      'dingbat',
+      'Numero Uno Zanahoria',
+    ]);
+  });
+
+  it('overrides user filter option with user-focus filter option if there is a conflict', () => {
```
Comment (robertknight) [inline src/sidebar/components/hooks/test/use-user-filter-options-test.js:134]: Where is the "user filter option" referred to here being set?

File: src/sidebar/components/hooks/use-user-filter-options.js (hunk 7 above)
Comment (robertknight) [reply src/sidebar/components/hooks/use-user-filter-options.js:14]: I think a hook is a natural choice here. Custom hooks are good for encapsulating the logic for fetching data from some data source and triggering a re-render of a UI component when that data source changes. The main decision point is what part of the logic goes in the hook and what goes in the calling component.

File: src/sidebar/components/hooks/use-user-filter-options.js [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+  const focusFilters = store.getFocusFilters();
+  const showDisplayNames = store.isFeatureEnabled('client_display_names');
+
+  return useMemo(() => {
+    // Determine unique users (authors) in annotation collection
+    const users = {};
+    annotations.forEach(annotation => {
+      // Display a username, not an account ID. If we're dealing with an
+      // account ID, we need to username() it
+      const _username = username(annotation.user)
```
Comment (robertknight) [inline src/sidebar/components/hooks/use-user-filter-options.js:26]: As a stylistic convention we tend to use a trailing rather than leading underscore when there is a need to disambiguate.

Since we're only showing saved annotations in the notebook, the `annotation.user` field should always be populated with a valid value and hence the ternary should be unnecessary here.

File: src/sidebar/components/hooks/use-user-filter-options.js (hunk 1 above)
Comment (robertknight) [reply src/sidebar/components/hooks/use-user-filter-options.js:36]: To check my understanding here, the practical use case for this currently is that if the teacher is in grading mode in the LMS and has a particular user focused, when they open the notebook they should see that particular user, even if that user has not created any annotations?

File: src/sidebar/components/hooks/use-user-filter-options.js (hunk 9 above)
Comment (robertknight) [inline src/sidebar/components/hooks/use-user-filter-options.js:46]: If my assumption about the validity of `annotation.user` above is correct, then the keys of `users` _will_ always be valid account IDs.

File: src/sidebar/components/hooks/use-user-filter-options.js [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+      users[maybeUsername] = focusFilters.user.display;
+    }
+
+    // Convert to `FilterOption` objects
+    const userOptions = Object.keys(users).map(user => {
+      return { display: users[user], value: user };
+    });
+
+    // TODO: Add current user as a filter option ("Me")
```
Comment (robertknight) [inline src/sidebar/components/hooks/use-user-filter-options.js:71]: If you're not planning to do this now, maybe remove this comment from this PR and come back to it later.

File: src/sidebar/components/hooks/use-user-filter-options.js [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+    userOptions.sort((a, b) =>
+      a.display.toLowerCase() > b.display.toLowerCase() ? 1 : -1
```
Comment (robertknight) [inline src/sidebar/components/hooks/use-user-filter-options.js:74]:
```suggestion
      a.display.localeCompare(b.display)
```

This compares the strings using the appropriate sorting comparison for the current locale, whereas using `<` just compares the UTF-16 code units in the string one by one.

File: src/sidebar/components/notebook-results.js [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+            ({visibleCount} {visibleCount === 1 ? 'annotation' : 'annotations'})
+          </em>
+        </span>
+      )}
+    </span>
+  );
+}
+
+NotebookResults.propTypes = {
+  loadAnnotationsService: propTypes.object,
```
Comment (robertknight) [inline src/sidebar/components/notebook-results.js:54]: This prop is not used by the component.

File: src/sidebar/components/notebook-results.js [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+ * - Annotations are filtered: "X results [(and Y more)]"
+ */
+function NotebookResults() {
+  const store = useStoreProxy();
+  const forcedVisibleCount = store.forcedVisibleAnnotations().length;
+  const hasAppliedFilter = store.hasAppliedFilter();
+  const rootThread = useRootThread();
+  const visibleCount = countVisible(rootThread);
+
+  const hasResults = rootThread.children?.length > 0;
```
Comment (robertknight) [inline src/sidebar/components/notebook-results.js:22]: According to the type definitions for `Thread`, the `children` field is not optional, so the `?` is not needed.

File: src/sidebar/components/notebook-results.js [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+
+  const hasForcedVisible = forcedVisibleCount > 0;
+  const matchCount = visibleCount - forcedVisibleCount;
+  const threadCount = rootThread.children?.length ?? 0;
```
Comment (robertknight) [inline src/sidebar/components/notebook-results.js:26]: As above, `children` is a non-optional field.

File: src/sidebar/components/notebook-results.js [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+import { countVisible } from '../util/thread';
+
+/**
+ * Render count of annotations (or filtered results) visible in the notebook view
+ * There are three possible overall states:
+ * - No results (regardless of whether annotations are filtered): "No results"
+ * - Annotations are unfiltered: "X threads (Y annotations)"
+ * - Annotations are filtered: "X results [(and Y more)]"
+ */
+function NotebookResults() {
```
Comment (robertknight) [inline src/sidebar/components/notebook-results.js:15]: This component name to me suggests that it displays the results of the current search/filter, whereas it only displays the counts. Perhaps rename to `NotebookResultCount`?

File: src/sidebar/components/select-filter.js (hunk 10 above)
Comment (robertknight) [inline src/sidebar/components/select-filter.js:28]: Is this component intended to be a a foundation generic select that we could use for other purposes?

Can I suggest that the name `FilterSelect` would be more consistent with the `<Entity><Widget Type>` naming scheme for other components (eg. `UserMenu`, `SidebarView`, `TagList`).

Comment (robertknight) [reply src/sidebar/components/select-filter.js:28]: Starting out with a menu and potentially replacing it with a different component in future sounds reasonable to me. We may want to consider unifying it with the component used to select users in the LMS grading bar in future - I haven't thought through whether that will be a good fit yet.

File: src/sidebar/components/select-filter.js [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+          key={filterOption.value}
+          isSelected={filterOption.value === selected.value}
+          label={filterOption.display}
+        />
+      ))}
+    </Menu>
+  );
+}
+
+SelectFilter.propTypes = {
```
Comment (robertknight) [inline src/sidebar/components/select-filter.js:60]: Aside from this PR, we stopped using `propTypes` in the LMS frontend a while ago after we completed all the typedefs for props. We could do the same for the client soon.

File: src/sidebar/components/select-filter.js [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+function SelectFilter({
+  defaultOption,
+  icon,
+  onSelect,
+  options,
+  selectedOption,
+  title,
+}) {
+  const filterOptions = [defaultOption, ...options];
+  const selected = selectedOption ?? defaultOption;
```
Comment (robertknight) [inline src/sidebar/components/select-filter.js:37]: Currently the default option is not treated any differently than the others, so the responsibility for combining it with `filterOptions` could be the responsibility of the parent component. However if we wanted to render/treat it differently in future then we'd need to pass it.

File: src/sidebar/components/test/notebook-filters-test.js [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+    fakeUseUserFilterOptions.returns([
+      { display: 'One User', value: 'oneuser' },
+    ]);
+
+    const wrapper = createComponent();
+
+    wrapper
+      .find('SelectFilter')
+      .props()
+      .onSelect({ display: 'One User', value: 'oneuser' });
```
Comment (robertknight) [inline src/sidebar/components/test/notebook-filters-test.js:81]: This works fine as `onSelect` doesn't trigger any state updates currently. If it did then this would need to be wrapped in an `act` call.

File: src/sidebar/store/modules/filters.js [hunk 25]
```diff
@@ -285,6 +300,7 @@ function hasAppliedFilter(state) {
  * @prop {(filterName: string) => FilterOption|undefined} getFilter
  * @prop {() => Object<string,FilterOption>} getFilters
  * @prop {() => Object<string,string>} getFilterValues
+ * @prop {() => Object<string,FilterOption>} getFocusFilters
```
Comment (robertknight) [inline src/sidebar/store/modules/filters.js:303]: On master it is no longer necessary to write out store module types manually. You'll need to rebase to resolve the conflict.

**Review by robertknight — CHANGES_REQUESTED** [review]
Overall this looks good and I was able to follow it quite easily. I had some requests for changes on various minor details and I also suggested slight name changes to some of the components to better fit the pattern of other components.

File: src/sidebar/components/hooks/test/use-user-filter-options-test.js (hunk 12 above)
Comment (lyzadanger) [reply src/sidebar/components/hooks/test/use-user-filter-options-test.js:4]: OK, I updated to no `default` export. I also renamed the module to `use-filter-options` because it's plausible we might end up putting more than one related hook here, too.

File: src/sidebar/components/notebook-results.js (hunk 21 above)
Comment (lyzadanger) [reply src/sidebar/components/notebook-results.js:15]: Done.

File: src/sidebar/components/test/notebook-view-test.js (hunk 3 above)
Comment (lyzadanger) [reply src/sidebar/components/test/notebook-view-test.js:67]: Adding a test.

File: src/sidebar/components/hooks/use-user-filter-options.js (hunk 8 above)
Comment (lyzadanger) [reply src/sidebar/components/hooks/use-user-filter-options.js:59]: Removing.

File: src/sidebar/components/hooks/use-user-filter-options.js (hunk 9 above)
Comment (lyzadanger) [reply src/sidebar/components/hooks/use-user-filter-options.js:46]: Renaming to `focusedUser`.

File: src/sidebar/components/hooks/test/use-user-filter-options-test.js (hunk 13 above)
Comment (lyzadanger) [reply src/sidebar/components/hooks/test/use-user-filter-options-test.js:54]: Aye, so I can. Changing.

File: src/sidebar/components/hooks/use-user-filter-options.js (hunk 1 above)
Comment (lyzadanger) [reply src/sidebar/components/hooks/use-user-filter-options.js:36]: Its purpose is manifold:

1. Make sure that a focused user shows up in the list, even if they have authored no annotations in the current set
2. Make sure the focused user only shows up _once_ in the list (don't duplicate)
3. Make sure the user filter correctly shows the focused user as selected (this is the trickiest piece to explain)

The first requirement is fairly straightforward—we need to make sure the focused user is included in the list at all.

Perhaps an example is in order, as this is darned complicated to explain.

So let's talk about 2 and 3:

Let's say the app is configured to focus on a user, and the config for focused user provide a username and a display name for that user of `bumble` and `Bumble Bee`, respectively. Let's also say the Notebook is opened and the set of annotations contains several that were authored by `acct:bumble@flowertown` and a `display_name` (gleaned from one of their annotation objects)of `Bumble Q Bee`.

The hook will first generate a list of users that have authored at least one annotation—that would put `acct:bumble@flowertown / Bumble Q Bee` in the list of user filter options.

We need to make sure that we don't end up with two filter options for one person— `bumble / Bumble Bee` vs. `acct:bumble@flowertown / Bumble Q Bee`. A little bit of fiddling is required because the focus-user value could be either a username or an accountID. So that's part of this, eliminating dupes. This is where I got to initially and it satisfies item 2 above...but then I realized a wrinkle.

To satisfy item 3, we need to be sure that, in the case of the focused user already being in the users-with-annotations list, that filter option is superseded (replaced) by the focused-user filter variant. This matters when the focused-user filter value is a `username` instead of an `accountId`, as it is in our example.

Upon opening the Notebook in our example case,  there is an applied user filter (because of the focused-user config) for `bumble / Bumble Bee`. If we didn't replace the user filter option with the focus-filter variant—if we left it as `acct:bumble@flowertown / Bumble Q Bee`—the user filter would not correctly recognize that we are filtering on this user because the value of the applied (user-focus) filter is `bumble`, not `acct:bumble@flowertown`.

I'm making the assumption that the best UX for instructors in the grading mode would be to see annotations initially filtered by their focused user. I mean, we could just decide to not honor user-focus in Notebook, but I think it's nicer this way.

This whole screaming complexity could be averted if we only accepted `accountID`s for user-focus config.

Comment (lyzadanger) [reply src/sidebar/components/hooks/use-user-filter-options.js:36]: It being this hard to explain is more than a little bit of a design smell.

File: src/sidebar/components/hooks/test/use-user-filter-options-test.js (hunk 14 above)
Comment (lyzadanger) [reply src/sidebar/components/hooks/test/use-user-filter-options-test.js:134]: Bah, it's really hard to phrase this concisely (see my very wordy explanation of what's going on here). What I'm trying to summarize is: "In the case where focused user is defined and a filter option for that user already exists because they've also authored an annotation in the current annotation set, prefer the focus-filter variant over the annotation-user one." Swear word. Help.

Comment (lyzadanger) [reply src/sidebar/components/hooks/test/use-user-filter-options-test.js:134]: For now: `it('resolves user duplicates in favor of the focused user')`

File: src/sidebar/components/hooks/use-user-filter-options.js (hunk 15 above)
Comment (lyzadanger) [reply src/sidebar/components/hooks/use-user-filter-options.js:26]: Addressed.

File: src/sidebar/components/hooks/use-user-filter-options.js (hunk 9 above)
Comment (lyzadanger) [reply src/sidebar/components/hooks/use-user-filter-options.js:46]: A'ight, I'll make that assumption...

File: src/sidebar/components/select-filter.js (hunk 23 above)
Comment (lyzadanger) [reply src/sidebar/components/select-filter.js:37]: May I request that we see how this feels when we add another filter?

File: src/sidebar/store/modules/filters.js [hunk 26]
```diff
@@ -169,10 +169,21 @@ function changeFocusModeUser(user) {
...
 function setFilter(filterName, filterOption) {
-  return {
-    type: actions.SET_FILTER,
-    filterName,
-    filterOption,
+  return (dispatch, getState) => {
+    // If there is a filter conflict with focusFilters, deactivate focus
+    // mode to prevent unintended collisions and let the new filter value
+    // take precedence.
+    if (getState().filters.focusFilters?.[filterName]) {
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/sidebar/store/modules/filters.js:176]: For some reason I'm having trouble wrapping my head around what you're envisioning, would you mind elaborating a little bit?

**Comment by lyzadanger** [comment]
@robertknight Due to the number of modules that got renamed as a result of integrating feedback, etc, I ended up rewriting commit history entirely. I hope this doesn't slow you down.

I addressed all of your feedback items except for one that I left unresolved that I had a confusion over. Thanks!

File: src/sidebar/store/modules/filters.js (hunk 26 above)
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/store/modules/filters.js:177]: If you want one action to incorporate the effects of another, then one option is to use a thunk with `dispatch` as you have done here. Another is to call one reducer from within another:

```diff
diff --git a/src/sidebar/store/modules/filters.js b/src/sidebar/store/modules/filters.js
index e36e7eff5..c47b6f5a8 100644
--- a/src/sidebar/store/modules/filters.js
+++ b/src/sidebar/store/modules/filters.js
@@ -119,6 +119,15 @@ const update = {
   },
 
   SET_FILTER: function (state, action) {
+    // If there is a filter conflict with focusFilters, deactivate focus
+    // mode to prevent unintended collisions and let the new filter value
+    // take precedence.
+    let newState = /** @type {typeof state} */ ({});
+    if (state.focusFilters?.[action.filterName]) {
+      // eslint-disable-next-line new-cap
+      newState = update.SET_FOCUS_MODE(state, { active: false });
+    }
+
     const updatedFilters = {
       ...state.filters,
       [action.filterName]: action.filterOption,
@@ -127,7 +136,9 @@ const update = {
     if (action.filterOption?.value === '') {
       delete updatedFilters[action.filterName];
     }
-    return { filters: updatedFilters };
+    newState.filters = updatedFilters;
+
+    return newState;
   },
 
   SET_FILTER_QUERY: function (state, action) {
@@ -170,22 +181,7 @@ function changeFocusModeUser(user) {
  * @param {FilterOption} filterOption
  */
 function setFilter(filterName, filterOption) {
-  return (dispatch, getState) => {
-    // If there is a filter conflict with focusFilters, deactivate focus
-    // mode to prevent unintended collisions and let the new filter value
-    // take precedence.
-    if (getState().filters.focusFilters?.[filterName]) {
-      dispatch({
-        type: actions.SET_FOCUS_MODE,
-        active: false,
-      });
-    }
-    dispatch({
-      type: actions.SET_FILTER,
-      filterName,
-      filterOption,
-    });
-  };
+  return { type: actions.SET_FILTER, filterName, filterOption };
 }
 
 /** Set the query used to filter displayed annotations. */
```

There isn't a significant advantage to changing the current code to the above in this case. I'm just noting that it would have been an option.

File: src/sidebar/components/hooks/use-filter-options.js [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+
+      // Remove any pre-existing filter option corresponding to this user
+      // before (re-)adding it. The value defined by the focused user config
+      // needs to take precedence. This is so that if user focus is active,
+      // a filter component will correctly recognize the focused user as selected.
+      // Not doing this can result in two filter options for the same user:
+      // one using a username and one using an account ID.
+      delete users[focusedUser];
+
+      users[maybeUsername] = focusFilters.user.display;
```
Comment (robertknight) [inline src/sidebar/components/hooks/use-filter-options.js:58]: So the `maybeUsername` key here can be either a username or an account ID whereas the `users[account.user]` line above will always be an account ID key. Since in the current usage of the Notebook we're only intending to display users from a single authority, can I suggest that we make all of the keys in the `users` may usernames, as the lowest common denominator.

File: src/sidebar/components/select-filter.js (hunk 23 above)
Comment (robertknight) [reply src/sidebar/components/select-filter.js]: Yes. I was just thinking aloud here.

File: src/sidebar/components/test/notebook-view-test.js (hunk 3 above)
Comment (robertknight) [reply src/sidebar/components/test/notebook-view-test.js:67]: I generally avoid tests for parts of the output which are completely static (ie. is not conditionally rendered and doesn't contain any dynamic properties).

File: src/sidebar/store/modules/test/filters-test.js [hunk 28]
```diff
@@ -74,6 +74,26 @@ describe('sidebar/store/modules/filters', () => {
...
+        // No conflict in focusFilters on `elephant`
+        store.setFilter('elephant', {
+          value: 'pink',
+          display: 'Pink Elephant',
+        });
+
+        assert.isTrue(store.focusState().active);
+
+        store.setFilter('user', { value: '', display: 'Everybody' });
+        assert.isFalse(store.focusState().active);
```
Comment (robertknight) [inline src/sidebar/store/modules/test/filters-test.js:94]: To check my understanding - the purpose of this is to avoid a situation where focus mode configures the client to show only annotations from user A and filters configure the client to show only annotations from user B? In that case no annotations could match both constraints.

File: src/styles/sidebar/components/notebook-filters.scss [hunk 29]
```diff
@@ -0,0 +1,21 @@
+@use "../../mixins/utils";
+@use "../../variables" as var;
+
+.select-filter {
```
Comment (robertknight) [inline src/styles/sidebar/components/notebook-filters.scss:4]: This name doesn't match the file name. Can you rename the files/classes to make sure that the CSS class name, Preact component name and filenames are all aligned.

**Review by robertknight — CHANGES_REQUESTED** [review]
This is almost ready to go. A few minor issues I highlighted:

- I suggest making the `value` property of user filter options always be usernames, to avoid needing to deal with a mix of account IDs and usernames.
- There is a mismatch between CSS class names, .scss file name and Preact component name for one component

Once the above are addressed this is good to merge.

File: src/sidebar/components/hooks/use-filter-options.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/sidebar/components/hooks/use-filter-options.js:58]: OK, I think that this is reasonable. I think that part of me was thinking the really-long-term—I didn't want to create barriers to a glorious multi-authority future—but there are bigger fish to fry than this.

File: src/sidebar/components/notebook-result-count.js [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+
+  const hasForcedVisible = forcedVisibleCount > 0;
+  const matchCount = visibleCount - forcedVisibleCount;
+  const threadCount = rootThread.children.length;
+
+  return (
+    <span className="notebook-result-count">
+      {!hasResults && <h2>No results</h2>}
+      {hasResults && hasAppliedFilter && (
+        <span>
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/sidebar/components/notebook-result-count.js:31]: Can you help me understand the tradeoffs of using a `<Fragment>` versus using a containing element?

File: src/sidebar/store/modules/test/filters-test.js (hunk 28 above)
Comment (lyzadanger) [reply src/sidebar/store/modules/test/filters-test.js:94]: The correctly-filtered annotations would show up (as filtering by either `username` or `accountId` "works") but the user option in the filter-select would not be correctly shown as selected (it would incorrectly show that "Everybody"'s annotations were showing).

File: src/sidebar/components/hooks/use-filter-options.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/sidebar/components/hooks/use-filter-options.js:58]: Refreshingly, this was surgically contained all within part of the relevant hook, which is encouraging.

File: src/styles/sidebar/components/notebook-filters.scss (hunk 29 above)
Comment (lyzadanger) [reply src/styles/sidebar/components/notebook-filters.scss:4]: Hand-wavy excuse about how this is a relic of ongoing refactoring while I was implementing, etc...

File: src/sidebar/store/modules/filters.js (hunk 26 above)
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/sidebar/store/modules/filters.js:177]: Admittedly did not know this was even possible, so this has value just as educational content :).

File: src/sidebar/components/notebook-result-count.js (hunk 30 above)
[the remark this answers is not in the document]
Comment (robertknight) [reply src/sidebar/components/notebook-result-count.js:32]: The main benefit is that it avoids adding clutter to the DOM tree that doesn't serve any purpose for semantic or styling purposes. It's only a small benefit in many cases though.

**Review by robertknight — APPROVED** [review]
